packages/kbn-openapi-generator/src/cli.ts

```diff
@@ -36,6 +36,11 @@ export function runCli() {
             type: 'boolean',
             default: false,
           })
+          .option('outputDir', {
+            describe: 'Output directory for generated types',
+            demandOption: false,
+            type: 'string',
+          })
```

Comment:
It seems clear to me that the API schemas as a common artifact should be separate from the (server) API routes and (public) client hooks, so I agree that picking output destinations for client api hooks and server route stubs will be good and necessary - but I think allowing a separate output directory for the generated schema files would still be extra complexity even if we need to allow output destinations for those other artifacts because each type of artifact will have different output destinations and likely different needs around the structure of the output destination.

It feels very natural to me to keep the schema yaml files right next to the corresponding generated TS files because they're so closely linked together, so it's easier to flip back and forth between yaml and .gen.ts to see the effects of changes, less configuration and script logic to maintain, and more standardization across plugins so devs can work across Kibana more easily.

Maybe @andrew-goldstein can share his perspective on why the yaml files belong in `server`?

Reply:
> Maybe @andrew-goldstein can share his perspective on why the yaml files belong in `server`?

In this use case, the yaml file (schema):
1) Is the canonical definition of the `capabilities` server route, because the `kbn-openapi-generator` takes a _schema-first_ approach, where the schema is the authoritative definition of the server route. (For example, the typescript type describing the shape of the response returned by the `capabilities` server route, is derived / generated from the schema.)
2) The (one and only) implementation of the `capabilities` server route is in the `elastic_assistant` plugin:  `x-pack/plugins/elastic_assistant/server/routes/capabilities/get_capabilities_route.ts`
3) In the originating PR, the schema for the `capabilities` route was (and still is) naturally defined in `x-pack/plugins/elastic_assistant/server/schemas/capabilities/get_capabilities_route.schema.yaml`
4) I made [this comment in the originating PR](https://github.com/elastic/kibana/pull/174317#discussion_r1447728413), requesting that the `GetCapabilitiesResponse` type, (generated from the schema), be moved to the `kbn-elastic-assistant-common` package, because the generated type is required by both the client `packages/kbn-elastic-assistant`, and the server implementation `plugins/elastic_assistant/server`

When viewed side-by-side, the paths in `2` and `3` above:

```
x-pack/plugins/elastic_assistant/server/routes/capabilities
x-pack/plugins/elastic_assistant/server/schemas/capabilities
```

are naturally co-located in `x-pack/plugins/elastic_assistant/server`, because they are (respectively):
- The one-and-only implementation of the `capabilities` server route
- The canonical schema definition of the `capabilities` server route

AFIK, the _only_ driver to _separate_ the implementation and schema above is an implementation detail: the `GetCapabilitiesResponse` type _derived_ from the schema is required by both the client (`packages/kbn-elastic-assistant`) and server (`plugins/elastic_assistant/server`).

In this use case, the `outputDir` feature enables the `server/routes` and `server/schemas` to be (naturally) co-located in the same plugin, even though implementation details require the derived / generated types to live in a separate (common) package.

Reply:
I would say that the main driver for separating the route implementation from the yaml schema is that the schema defines the contract between API consumers and the API route while the API route defines only the server side portion of the API. In that sense, both the public and server code depend on the contract defined by the yaml schema, and that shows in the flow of dependencies: schema (yaml) -> generated schema (TS) -> API client/API route. If the yaml schema resides in `server`, this means that the API client in `public` ends up with a transitive dependency on files in `server` which I don't think is good.

> AFIK, the only driver to separate the implementation and schema above is an implementation detail: the GetCapabilitiesResponse type derived from the schema is required by both the client (packages/kbn-elastic-assistant) and server (plugins/elastic_assistant/server).

After writing the first paragraph above and comparing to what you wrote here, they're basically the same idea but with different conclusions. IMO if we're deriving a type from the yaml spec and then using that type in `public` and `server`, then the yaml spec must go in `common` to avoid `public` having a dependency on `server`. I would think that, as a rule, if every `server` folder were deleted we should still be able to build the client application with only the files in `public` and `common`. Also note that for most API routes, there's both a request and response type needed by the client.

If we're going to separate the OpenAPI specs from the generated zod schemas, to me it would make more sense to flip it around and put the yaml specs in `common` and put the generated TS files in `server`, as the public client does not care about the server side validation done by the schemas - it only cares about the HTTP API contract. As you mentioned, right now the public client does need the types from the generated zod schemas, so the zod schemas need to go in common as well. However, one of the additional goals we have for generation capabilities from OpenAPI specs is generating TS types directly from yaml, without having to use things like `z.infer`. This can help avoid some implementation detail issues, such as TS type inference failing due to depth and difficulty with defining recursive schemas. This could also help here by allowing us to generate TS types for the API contract in `common` but put the zod runtime validation schemas in `server`. Even with this future capability, though, if the OpenAPI schemas are in `server` then `public` ends up with a transitive dependency on `server`.

Let me know if you want to zoom to discuss this synchronously!

Reply:
Moving the schema to `common` is reasonable. Also happy to discuss synchronously @marshallmain

Reply:
`outputDir` removed in [4122a12](https://github.com/elastic/kibana/pull/174718/commits/4122a12f9eedf7cb97210420562c8131d57f244d), and `*.schema.yaml` files (and corresponding scripts) moved to `kbn-elastic-assistant-common` package.